### ISSUE TITLE
tcpdump bootp dissector not print verdor-specific attributes if sname is not NULL

### ISSUE THREAD

**Issue by kondratev** [the-tcpdump-group/tcpdump#685]
[dhcp.pcap.gz](https://github.com/the-tcpdump-group/tcpdump/files/2110985/dhcp.pcap.gz)

```diff --git a/print-bootp.c b/print-bootp.c
index 8d2166c2..62b337cb 100644
--- a/print-bootp.c
+++ b/print-bootp.c
@@ -361,7 +361,7 @@ bootp_print(netdissect_options *ndo,
        if (EXTRACT_U_1(bp->bp_sname)) {
                ND_PRINT("\n\t  sname \"");
                if (nd_printztn(ndo, bp->bp_sname, (u_int)sizeof(bp->bp_sname),
-                               ndo->ndo_snapend)) {
+                               ndo->ndo_snapend) == 0) {
                        ND_PRINT("\"");
                        nd_print_trunc(ndo);
                        return;
@@ -372,7 +372,7 @@ bootp_print(netdissect_options *ndo,
        if (EXTRACT_U_1(bp->bp_file)) {
                ND_PRINT("\n\t  file \"");
                if (nd_printztn(ndo, bp->bp_file, (u_int)sizeof(bp->bp_file),
-                               ndo->ndo_snapend)) {
+                               ndo->ndo_snapend) == 0) {
                        ND_PRINT("\"");
                        nd_print_trunc(ndo);
                        return;
```

**Comment by infrastation** [comment]
backtick-backtick-backtick-d-i-f-f
(diff contents)
backtick-backtick-backtick

will format your comment the way you probably intended (or open a pull request instead).

**Comment by infrastation** [comment]
It would also help to describe how the bug reproduces in the wild.

**Comment by kondratev** [comment]
@infrastation attached file must print 

08:35:16.118314 IP (tos 0x0, ttl 64, id 18509, offset 0, flags [none], proto UDP (17), length 306)
    10.1.43.1.67 > 31.192.168.162.68: [udp sum ok] BOOTP/DHCP, Reply, length 278, xid 0x1fa8280a, Flags [none] (0x0000)
          Your-IP 31.192.168.162
          Client-Ethernet-Address 00:15:60:55:0f:be
          sname "10.0.2.6"
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            Subnet-Mask Option 1, length 4: 255.255.255.248
            Lease-Time Option 51, length 4: 300
            Default-Gateway Option 3, length 4: 31.192.190.9
            DHCP-Message Option 53, length 1: ACK
            Domain-Name-Server Option 6, length 8: 87.224.213.1,87.224.197.1
            Server-ID Option 54, length 4: 10.0.2.6
        0x0000:  0015 6055 0fbe 3c61 042a 07ca 0800 4500 ..`U..<a.*....E.
        0x0010:  0132 484d 0000 4011 340a 0a01 2b01 1fc0 .2HM..@.4...+...
        0x0020:  a8a2 0043 0044 011e dd4c 0201 0600 1fa8 ...C.D...L......
        0x0030:  280a 0000 0000 0000 0000 1fc0 a8a2 0000 (...............
        0x0040:  0000 0000 0000 0015 6055 0fbe 0000 0000 ........`U......
        0x0050:  0000 0000 0000 3130 2e30 2e32 2e36 0000 ......10.0.2.6..
        0x0060:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0070:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0080:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0090:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00a0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00b0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00c0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00d0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00e0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00f0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0100:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0110:  0000 0000 0000 6382 5363 0104 ffff fff8 ......c.Sc......
        0x0120:  3304 0000 012c 0304 1fc0 be09 3501 0506 3....,......5...
        0x0130:  0857 e0d5 0157 e0c5 0136 040a 0002 06ff .W...W...6......
but without this fix prints
08:35:16.118314 IP (tos 0x0, ttl 64, id 18509, offset 0, flags [none], proto UDP (17), length 306)
    10.1.43.1.67 > 31.192.168.162.68: [udp sum ok] BOOTP/DHCP, Reply, length 278, xid 0x1fa8280a, Flags [none] (0x0000)
          Your-IP 31.192.168.162
          Client-Ethernet-Address 00:15:60:55:0f:be
          sname "10.0.2.6"[|bootp]
        0x0000:  0015 6055 0fbe 3c61 042a 07ca 0800 4500 ..`U..<a.*....E.
        0x0010:  0132 484d 0000 4011 340a 0a01 2b01 1fc0 .2HM..@.4...+...
        0x0020:  a8a2 0043 0044 011e dd4c 0201 0600 1fa8 ...C.D...L......
        0x0030:  280a 0000 0000 0000 0000 1fc0 a8a2 0000 (...............
        0x0040:  0000 0000 0000 0015 6055 0fbe 0000 0000 ........`U......
        0x0050:  0000 0000 0000 3130 2e30 2e32 2e36 0000 ......10.0.2.6..
        0x0060:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0070:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0080:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0090:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00a0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00b0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00c0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00d0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00e0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x00f0:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0100:  0000 0000 0000 0000 0000 0000 0000 0000 ................
        0x0110:  0000 0000 0000 6382 5363 0104 ffff fff8 ......c.Sc......
        0x0120:  3304 0000 012c 0304 1fc0 be09 3501 0506 3....,......5...
        0x0130:  0857 e0d5 0157 e0c5 0136 040a 0002 06ff .W...W...6......

**Comment by fxlb** [comment]
@kondratev Could you attach the pcap test file in a zip archive ?

**Comment by guyharris** [comment]
To quote the comment at the beginning of `nd_printztn()`:

     * Return the number of bytes of string processed, including the
     * terminating null, if not truncated.  Return 0 if truncated.

So that indicates that checking whether the return value is zero, rather than whether it's non-zero, would be the correct thing to do; i.e. the change proposed is correct.

However, it also indicates that, *regardless* of whether the string is null or non-null, `nd_printztn()` should return a non-zero value, because a null string contains a terminating NUL, and therefore, as the number of bytes of string processed includes the terminating NUL, it'd return 1 for an empty string, so the code won't print anything after the sname *regardless* of whether it's null or non-null - it'll only print stuff after the sname if even the full sname isn't present in the captured data (i.e., it'll only print stuff after the sname if there isn't even a full sname, much less anything after the sname).

See, for example, the use of `nd_printztn()` in the TFTP dissector:

                ND_PRINT("\"");
                ui = nd_printztn(ndo, bp, length, ndo->ndo_snapend);
                ND_PRINT("\"");
                if (ui == 0)
                        goto trunc;

1.8.1 prints the second packet in the capture in question correctly - it does

	ND_TCHECK2(bp->bp_sname[0], 1);		/* check first char only */
	if (*bp->bp_sname) {
		ND_PRINT((ndo, "\n\t  sname \""));
		if (fn_print(ndo, bp->bp_sname, ndo->ndo_snapend)) {
			ND_PRINT((ndo, "\""));
			ND_PRINT((ndo, "%s", tstr + 1));
			return;
		}
		ND_PRINT((ndo, "\""));
	}

and `fn_print()` "Return[s] true if truncated.", i.e. returns a non-zero value.

The problem may be that `nd_printztn()` returns a *zero* value when truncated, but the test was left checking treating a *non-zero* return value as an indication that the field was truncated.

I've checked this fix in as 202ad447124735307fcb08d3a724a8f6c80c3290.

**Comment by guyharris** [comment]
> 1.8.1 prints the second packet in the capture in question correctly

Sorry, that's *4*.8.x prints the second packet in the capture in question correctly; 4.9.x doesn't.
